# **Graded Challenge 3**

# **Introduction**


Nama    : Fitri Octaviani

Batch   : HCK-006

# **Problem Statement**

Analisis data berdasarkan trend transaksi, trend users, dan trend pendapatan pada platform e-commerce dalam kurun waktu Q1 sampai Q3 pada tahun 2022, dengan tujuan untuk mengevaluasi kinerja penjualan berdasarkan trend yang mempengaruhi pertumbuhan atau penurunan aktivitas penjualan.

# **Data Loading**

In [ ]:
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id = "phase-0-390008" #GUNAKAN GCP PROJECT-ID KALIAN MASING-MASING
client = bigquery.Client(project=project_id)

Authenticated


In [ ]:
# melihat data dari distribution_centers
distribution_centers = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.distribution_centers`

''').to_dataframe()
distribution_centers

id                                         name  latitude  longitude
0   1                                   Memphis TN   35.1174   -89.9711
1   2                                   Chicago IL   41.8369   -87.6847
2   3                                   Houston TX   29.7604   -95.3698
3   4                               Los Angeles CA   34.0500  -118.2500
4   5                               New Orleans LA   29.9500   -90.0667
5   6  Port Authority of New York/New Jersey NY/NJ   40.6340   -73.7834
6   7                              Philadelphia PA   39.9500   -75.1667
7   8                                    Mobile AL   30.6944   -88.0431
8   9                                Charleston SC   32.7833   -79.9333
9  10                                  Savannah GA   32.0167   -81.1167

In [ ]:
# melihat data dari events
events = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.events`

''').to_dataframe()
events

id  user_id  sequence_number  \
0        1621464     <NA>                2   
1        2005888     <NA>                3   
2         564903    43192                6   
3        1366402     <NA>                3   
4         900150    68969                6   
...          ...      ...              ...   
2436640   799759    61241                1   
2436641  1646893     <NA>                1   
2436642   843133    64539                4   
2436643   843156    64539                1   
2436644   843152    64539               10   

                                   session_id                created_at  \
0        45acfb03-4949-47a7-9a21-14945ee5e927 2019-04-04 12:28:00+00:00   
1        90a5ed30-0382-41e0-ad57-82b708820bcf 2021-06-26 15:06:00+00:00   
2        6b91a904-062a-4493-bbc6-9633c8f88ab0 2021-07-01 11:57:22+00:00   
3        9429966f-bd9e-4ce9-9c3f-6769769e0d14 2020-07-13 03:01:00+00:00   
4        64f0aa2c-c179-4854-b8fc-84aca2304b28 2023-04-28 15:08:43+00:00   
...                                       ...                       ...   
2436640  8df82be6-2032-4f3c-b9a1-423332327a23 2022-11-18 08:56:05+00:00   
2436641  2172e282-1d75-49fa-b140-58f871ae8af5 2019-07-15 04:54:00+00:00   
2436642  391ff296-0a8c-4993-9b90-67fb5d72fc55 2023-04-22 06:34:03+00:00   
2436643  c6931bb1-09ca-4a5f-ac9d-1d663ec74a0e 2023-04-22 07:25:15+00:00   
2436644  6799d053-d4b9-48b2-b388-7af80e046456 2023-04-22 05:12:53+00:00   

              ip_address                 city              state postal_code  \
0          27.152.228.22            Bogatynia       Dolnośląskie          59   
1         138.183.243.51            Bogatynia       Dolnośląskie          59   
2          90.88.230.150            Bogatynia       Dolnośląskie          59   
3          158.38.251.11            Bogatynia       Dolnośląskie          59   
4           72.223.115.8  Gorzów Wielkopolski           Lubuskie          66   
...                  ...                  ...                ...         ...   
2436640   154.29.123.165              Panambi  Rio Grande do Sul   98280-000   
2436641  156.250.139.146            Carazinho  Rio Grande do Sul   99500-000   
2436642    180.169.52.74            Carazinho  Rio Grande do Sul   99500-000   
2436643     198.59.92.19            Carazinho  Rio Grande do Sul   99500-000   
2436644   169.87.119.102            Carazinho  Rio Grande do Sul   99500-000   

         browser traffic_source  \
0         Chrome       Facebook   
1         Chrome       Facebook   
2         Chrome          Email   
3             IE        Adwords   
4        Firefox        Adwords   
...          ...            ...   
2436640   Safari          Email   
2436641   Chrome        Adwords   
2436642   Safari          Email   
2436643   Chrome        Organic   
2436644   Chrome          Email   

                                                       uri  event_type  
0                                                    /cart        cart  
1                                                    /cart        cart  
2                                                    /cart        cart  
3                                                    /cart        cart  
4                                                    /cart        cart  
...                                                    ...         ...  
2436640  /department/women/category/shorts/brand/7foral...  department  
2436641  /department/women/category/maternity/brand/mot...  department  
2436642  /department/men/category/fashionhoodies&sweats...  department  
2436643  /department/men/category/sleep&lounge/brand/ma...  department  
2436644       /department/men/category/shorts/brand/savane  department  

[2436645 rows x 13 columns]

In [ ]:
# melihat data dari inventory_items
inventory_items = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.inventory_items`

''').to_dataframe()
inventory_items

id  product_id                created_at  \
0         6089       16898 2022-01-12 16:31:10+00:00   
1         6090       16898 2021-07-29 17:31:00+00:00   
2        62802       16898 2022-02-18 00:04:32+00:00   
3        62803       16898 2020-09-17 01:28:00+00:00   
4        62804       16898 2021-09-01 06:55:00+00:00   
...        ...         ...                       ...   
493259  404211       14392 2020-04-12 18:39:00+00:00   
493260  404212       14392 2020-02-13 09:20:00+00:00   
493261  478201       14392 2022-02-25 14:04:47+00:00   
493262  478202       14392 2020-01-31 06:30:00+00:00   
493263  478203       14392 2021-06-05 10:50:00+00:00   

                         sold_at       cost product_category  \
0      2022-02-14 02:10:10+00:00  13.925000      Tops & Tees   
1                            NaT  13.925000      Tops & Tees   
2      2022-03-25 12:36:32+00:00  13.925000      Tops & Tees   
3                            NaT  13.925000      Tops & Tees   
4                            NaT  13.925000      Tops & Tees   
...                          ...        ...              ...   
493259                       NaT  15.752521        Maternity   
493260                       NaT  15.752521        Maternity   
493261 2022-04-05 09:34:47+00:00  15.752521        Maternity   
493262                       NaT  15.752521        Maternity   
493263                       NaT  15.752521        Maternity   

                                             product_name  \
0                   Quiksilver Waterman Men's On The Rise   
1                   Quiksilver Waterman Men's On The Rise   
2                   Quiksilver Waterman Men's On The Rise   
3                   Quiksilver Waterman Men's On The Rise   
4                   Quiksilver Waterman Men's On The Rise   
...                                                   ...   
493259  La Leche League International Pullover Nursing...   
493260  La Leche League International Pullover Nursing...   
493261  La Leche League International Pullover Nursing...   
493262  La Leche League International Pullover Nursing...   
493263  La Leche League International Pullover Nursing...   

                                    product_brand  product_retail_price  \
0                                                             25.000000   
1                                                             25.000000   
2                                                             25.000000   
3                                                             25.000000   
4                                                             25.000000   
...                                           ...                   ...   
493259  La Leche League Sports Nurising Bra White             35.720001   
493260  La Leche League Sports Nurising Bra White             35.720001   
493261  La Leche League Sports Nurising Bra White             35.720001   
493262  La Leche League Sports Nurising Bra White             35.720001   
493263  La Leche League Sports Nurising Bra White             35.720001   

       product_department                       product_sku  \
0                     Men  22811EE19846217512507785E74D12CC   
1                     Men  22811EE19846217512507785E74D12CC   
2                     Men  22811EE19846217512507785E74D12CC   
3                     Men  22811EE19846217512507785E74D12CC   
4                     Men  22811EE19846217512507785E74D12CC   
...                   ...                               ...   
493259              Women  625D2D3101C23C47A46984741C232FED   
493260              Women  625D2D3101C23C47A46984741C232FED   
493261              Women  625D2D3101C23C47A46984741C232FED   
493262              Women  625D2D3101C23C47A46984741C232FED   
493263              Women  625D2D3101C23C47A46984741C232FED   

        product_distribution_center_id  
0                                    3  
1                                    3  
2                                    3  
3              

In [ ]:
# melihat data dari order_items
order_items = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.order_items`

''').to_dataframe()
order_items

id  order_id  user_id  product_id  inventory_item_id      status  \
0        25653     17657    13887       13606              69258     Shipped   
1        40254     27783    21893       13606             108683     Shipped   
2        83516     57720    45714       13606             225837     Shipped   
3        32660     22537    17814       13606              88240    Complete   
4        24216     16670    13092       13606              65386  Processing   
...        ...       ...      ...         ...                ...         ...   
182660   97910     67599    53627        3449             264733     Shipped   
182661   44704     30906    24382        3449             120777    Complete   
182662  138472     95558    76008        3449             374014    Complete   
182663  160178    110351    87817        3449             432670    Complete   
182664   74189     51294    40632        3449             200652  Processing   

                      created_at                shipped_at  \
0      2023-04-15 22:27:44+00:00 2023-04-17 12:53:00+00:00   
1      2021-09-28 01:11:25+00:00 2021-09-26 12:23:00+00:00   
2      2023-06-16 03:57:35+00:00 2023-06-16 22:26:00+00:00   
3      2022-05-17 05:05:40+00:00 2022-05-20 04:01:00+00:00   
4      2023-05-29 05:59:55+00:00                       NaT   
...                          ...                       ...   
182660 2023-06-15 02:58:02+00:00 2023-06-18 02:38:00+00:00   
182661 2023-01-08 12:29:16+00:00 2023-01-07 14:50:00+00:00   
182662 2021-06-15 17:36:14+00:00 2021-06-16 12:59:00+00:00   
182663 2021-02-18 05:54:34+00:00 2021-02-18 13:01:00+00:00   
182664 2020-11-04 17:40:47+00:00                       NaT   

                    delivered_at returned_at  sale_price  
0                            NaT         NaT        2.50  
1                            NaT         NaT        2.50  
2                            NaT         NaT        2.50  
3      2022-05-23 03:06:00+00:00         NaT        2.50  
4                            NaT         NaT        2.50  
...                          ...         ...         ...  
182660                       NaT         NaT        9.82  
182661 2023-01-11 14:30:00+00:00         NaT        9.82  
182662 2021-06-19 23:22:00+00:00         NaT        9.82  
182663 2021-02-19 11:19:00+00:00         NaT        9.82  
182664                       NaT         NaT        9.82  

[182665 rows x 11 columns]

In [ ]:
# melihat data dari orders
orders = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.orders`

''').to_dataframe()
orders

order_id  user_id      status gender                       created_at  \
0             13       13     Shipped      F        2021-11-16 13:28:00+00:00   
1             19       18     Shipped      F 2023-06-16 03:01:31.810245+00:00   
2             21       19     Shipped      F        2022-09-15 07:07:00+00:00   
3             23       21     Shipped      F        2021-11-06 14:44:00+00:00   
4             24       22     Shipped      F        2023-05-27 06:51:00+00:00   
...          ...      ...         ...    ...                              ...   
125807    125717    99929  Processing      M        2023-04-25 00:47:00+00:00   
125808    125731    99938  Processing      M        2023-03-23 06:28:00+00:00   
125809    125736    99943  Processing      M        2021-12-12 00:14:00+00:00   
125810    125743    99948  Processing      M        2023-05-09 04:04:00+00:00   
125811    125772    99972  Processing      M        2022-01-28 09:24:00+00:00   

       returned_at                       shipped_at delivered_at  num_of_item  
0              NaT        2021-11-17 22:00:00+00:00          NaT            1  
1              NaT 2023-06-16 04:52:31.810245+00:00          NaT            1  
2              NaT        2022-09-16 16:59:00+00:00          NaT            1  
3              NaT        2021-11-08 23:41:00+00:00          NaT            1  
4              NaT        2023-05-28 19:04:00+00:00          NaT            1  
...            ...                              ...          ...          ...  
125807         NaT                              NaT          NaT            1  
125808         NaT                              NaT          NaT            1  
125809         NaT                              NaT          NaT            3  
125810         NaT                              NaT          NaT            1  
125811         NaT                              NaT          NaT            2  

[125812 rows x 9 columns]

In [ ]:
# melihat data dari products
products = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.products`

''').to_dataframe()
products

id       cost                       category  \
0      27569  92.652563                           Swim   
1      27445  24.719661                           Swim   
2      27457  15.897600                           Swim   
3      27466  17.850000                           Swim   
4      27481  29.408001                           Swim   
...      ...        ...                            ...   
29115   1993  22.150999  Fashion Hoodies & Sweatshirts   
29116   2062  18.973360  Fashion Hoodies & Sweatshirts   
29117   2308   9.028020  Fashion Hoodies & Sweatshirts   
29118   1852  17.674240  Fashion Hoodies & Sweatshirts   
29119   2296  17.533261  Fashion Hoodies & Sweatshirts   

                                                    name  \
0         2XU Men's Swimmers Compression Long Sleeve Top   
1             TYR Sport Men's Square Leg Short Swim Suit   
2        TYR Sport Men's Solid Durafast Jammer Swim Suit   
3      TYR Sport Men's Swim Short/Resistance Short Sw...   
4                        TYR Alliance Team Splice Jammer   
...                                                  ...   
29115  Comfort Colors C1598 Womens Full-Zip Hooded Fl...   
29116  Comfort Colors Pigment-Dyed Ladies Full-Zip Ho...   
29117  Comfort Colors by Chouinard Women's Drawstring...   
29118  DV by Dolce Vita Women's Larisa Hooded Sweatshirt   
29119    Independent Trading Co Deluxe Fleece Zip Hoodie   

                        brand  retail_price department  \
0                         2XU    150.410004        Men   
1                         TYR     38.990002        Men   
2                         TYR     27.600000        Men   
3                         TYR     30.000000        Men   
4                         TYR     45.950001        Men   
...                       ...           ...        ...   
29115          Comfort Colors     52.119999      Women   
29116          Comfort Colors     39.860001      Women   
29117          Comfort Colors     18.020000      Women   
29118        DV by Dolce Vita     34.520000      Women   
29119  Independent Trading Co     36.990002      Women   

                                    sku  distribution_center_id  
0      B23C5765E165D83AA924FA8F13C05F25                       1  
1      2AB7D3B23574C3DEA2BD278AFD0939AB                       1  
2      8F831227B0EB6C6D09A0555531365933                       1  
3      67317D6DCC4CB778AEB9219565F5456B                       1  
4      213C888198806EF1A0E2BBF2F4855C6C                       1  
...                                 ...                     ...  
29115  C5A4E7E6882845EA7BB4D9462868219B                      10  
29116  2D405B367158E3F12D7C1E31A96B3AF3                      10  
29117  E58AEA67B01FA747687F038DFDE066F6                      10  
29118  EB1E78328C46506B46A4AC4A1E378B91                      10  
29119  130F1A8E9E102707F3F91B010F151B0B                      10  

[29120 rows x 9 columns]

In [ ]:
# melihat data dari users
users = client.query('''
SELECT *
FROM `bigquery-public-data.thelook_ecommerce.users`

''').to_dataframe()
users

id   first_name   last_name                            email  age  \
0      22183  Christopher     Jackson   christopherjackson@example.com   54   
1      95469         Lori    Cardenas         loricardenas@example.net   39   
2      25897        Alvin     Parsons         alvinparsons@example.net   70   
3      25928      Richard     Johnson       richardjohnson@example.com   63   
4      50770    Alejandra       Hardy       alejandrahardy@example.com   24   
...      ...          ...         ...                              ...  ...   
99995  88306        Kevin     Jackson         kevinjackson@example.net   15   
99996   6241         John      Barber           johnbarber@example.com   61   
99997  14772       Curtis      Guzman         curtisguzman@example.com   32   
99998  90014         Sean       Ellis            seanellis@example.com   39   
99999  97099    Catherine  Fitzgerald  catherinefitzgerald@example.org   68   

      gender                                 state  \
0          M                                   Mie   
1          F                                   Mie   
2          M                                  Acre   
3          M                                  Acre   
4          F                                  Acre   
...      ...                                   ...   
99995      M  Jeju Special Self-Governing Province   
99996      M  Jeju Special Self-Governing Province   
99997      M  Jeju Special Self-Governing Province   
99998      M  Jeju Special Self-Governing Province   
99999      F  Jeju Special Self-Governing Province   

                       street_address postal_code           city      country  \
0      16270 Jonathan Forges Apt. 180    513-0836    Suzuka City        Japan   
1                 11827 Justin Plains    513-0836    Suzuka City        Japan   
2        53983 Murphy Brook Suite 044   69917-400     Rio Branco       Brasil   
3                   6836 Campos Falls   69917-400     Rio Branco       Brasil   
4           807 Gordon Isle Suite 902   69917-400     Rio Branco       Brasil   
...                               ...         ...            ...          ...   
99995                 271 Carson View     699-930  Seogwipo City  South Korea   
99996              622 Sherman Circle     699-940  Seogwipo City  South Korea   
99997               25024 Greg Ridges     699-940  Seogwipo City  South Korea   
99998        293 James Trace Apt. 802     699-940  Seogwipo City  South Korea   
99999                  5402 Jack Burg     699-940  Seogwipo City  South Korea   

        latitude   longitude traffic_source                created_at  
0      34.851814  136.508713        Display 2022-06-14 11:24:00+00:00  
1      34.851814  136.508713          Email 2020-01-10 09:44:00+00:00  
2      -9.945568  -67.835610       Facebook 2021-01-26 17:54:00+00:00  
3      -9.945568  -67.835610         Search 2022-09-29 16:55:00+00:00  
4      -9.945568  -67.835610        Organic 2021-05-06 13:17:00+00:00  
...          ...         ...            ...                       ...  
99995  33.220465  126.237860         Search 2019-07-13 09:21:00+00:00  
99996  33.305744  126.675429        Organic 2022-05-01 14:12:00+00:00  
99997  33.305744  126.675429        Organic 2020-06-10 05:20:00+00:00  
99998  33.305744  126.675429        Organic 2019-06-02 09:25:00+00:00  
99999  33.305744  126.675429         Search 2022-10-17 01:58:00+00:00  

[100000 rows x 15 columns]

## **SOAL 1**

> Berapa jumlah transaksi yang berstatus Complete tiap bulan selama Q1 sampai Q3 di tahun 2022?

In [ ]:
# menghitung jumlah transaksi
Jumlah_transaksi = client.query('''
SELECT
    # extract tahun, bulan, dan quarter dari kolom created_at
    EXTRACT (year FROM created_at) as Tahun,
    EXTRACT (month FROM created_at) as Bulan,
    EXTRACT (QUARTER FROM created_at) as Quarter,
    # menghitung jumlah transaksi berdasarkan status
    COUNT (status) as Jumlah_Transaksi
FROM bigquery-public-data.thelook_ecommerce.order_items
WHERE
    # Kondisi dimana status yang diambil adalah Complete
    status = 'Complete'
    # Tahun 2022, Bulan 1 sampai 9, Quarter kurang dari 4
    AND EXTRACT (year FROM created_at) = 2022
    AND EXTRACT (month FROM created_at) <= 9
    AND EXTRACT(QUARTER FROM created_at) < 4
# mengelompokkan baris
GROUP BY Tahun, Bulan, Quarter
ORDER BY Bulan

''').to_dataframe()
Jumlah_transaksi

Tahun  Bulan  Quarter  Jumlah_Transaksi
0   2022      1        1               917
1   2022      2        1               902
2   2022      3        1              1105
3   2022      4        2              1053
4   2022      5        2              1132
5   2022      6        2              1334
6   2022      7        3              1384
7   2022      8        3              1386
8   2022      9        3              1483

**INSIGHT**

Data yang memiliki jumlah transaksi yang berstatus Complete tersedia pada tabel order items. Untuk menghitung status Complete tersebut perlu menggunakan aggregasi COUNT, dan quartil yang digunakan adalah kolom created_at menggunakan fungsi EXTRACT untuk memisahkan Q1 sampai Q3 dengan menggunakan data bigquery-public-data.thelook_ecommerce.order_items.

- Jumlah total pengguna yang berstatus complete selama 3 quartil adalah 10.696 users.

- Jumlah total pengguna terendah tercatat pada bulan ke-2 sebesar 902 users, sedangkan jumlah pengguna tertinggi terjadi pada bulan ke-9 sebanyak 1483.

- Jumlah transaksi yang berstatus Complete semakin meningkat dari tiap bulan atau quartalnya, hal ini merupakan indikator positif untuk bisnis dan dapat lebih ditingkatkan lagi agar mencapai targat yang diinginkan.

## **SOAL 2**

> Berikan informasi total penjualan (dalam USD) tiap bulan selama Q1 sampai Q3 di tahun 2022! (Hanya yang transaksi berstatus Complete).







In [ ]:
# Menghitung total penjualan tiap bulan
total_penjualan = client.query('''
SELECT
  # extract tahun, bulan, dan quarter dari kolom shipped_at
  EXTRACT(YEAR FROM shipped_at) as Tahun,
  EXTRACT(MONTH FROM shipped_at) as Bulan,
  EXTRACT (QUARTER FROM shipped_at) as Quarter,
  # melakukan perhitungan pada sekumpulan nilai dari product_retail_price
  SUM(product_retail_price) AS Total_penjualan
FROM
  `bigquery-public-data.thelook_ecommerce.order_items` AS items
JOIN
  `bigquery-public-data.thelook_ecommerce.inventory_items` AS inventory
# spesifikasi kondisi join dari dua dataset
ON
  items.product_id = inventory.product_id
WHERE
  status = 'Complete'
  AND EXTRACT(YEAR FROM shipped_at) = 2022
  AND EXTRACT(MONTH FROM shipped_at) BETWEEN 1 AND 9
  AND EXTRACT(QUARTER FROM shipped_at) < 4
GROUP BY Tahun, Bulan, Quarter
ORDER BY Bulan

''').to_dataframe()
total_penjualan

Tahun  Bulan  Quarter  Total_penjualan
0   2022      1        1     1.179685e+06
1   2022      2        1     1.117165e+06
2   2022      3        1     1.301672e+06
3   2022      4        2     1.271810e+06
4   2022      5        2     1.434485e+06
5   2022      6        2     1.555542e+06
6   2022      7        3     1.558493e+06
7   2022      8        3     1.655541e+06
8   2022      9        3     1.775233e+06

**INSIGHT**

Untuk menghitung total penjualan tiap bulan diperlukan 2 dataset dengan menggunakan fungsi JOIN, order_items dan inventory_items. Setelah itu menggunakan aggregasi SUM untuk menghitung total penjualan tiap bulan pada product_retail_price dengan status complete pada Q1 sampai Q3 tahun 2022.

- Total pendapatan selama 3 quartil adalah 12849626 USD.

- Pendapatan terendah tercatat pada bulan ke-2 sebesar 1.117165 USD, sementara pendapatan tertinggi terjadi pada bulan ke-9 sebanyak 1.775233 USD.

- Dikarenakan pada bulan ke-dua dan ke-empat sempat mengalami penurunan jumlah transaksi yang berstatus Complete, sehingga total penjualan dalam USD yang diperoleh dari bulan ke-dua dan ke-empat pun mengalami penurunan dalam total penjualan (USD). Akan tetapi untuk bulan selanjutnya mengalami peningkatan dalam total penjualan (USD).

## **SOAL 3**

> Berapa user yang melakukan transaksi berstatus Complete di tiap bulan dari Q1 sampai Q3 2022?

In [ ]:
# menghitung user yang melakukan transaksi di tiap bulan
user_transaksi = client.query('''
SELECT
    EXTRACT (year FROM items.created_at) as Tahun,
    EXTRACT (month FROM items.created_at) as Bulan,
    EXTRACT (quarter FROM items.created_at) as Quarter,
    # hanya values yang berbeda agar data users yang sama tidak berulang kali
    COUNT (DISTINCT user_id) as Users
FROM `bigquery-public-data.thelook_ecommerce.users` as users
JOIN `bigquery-public-data.thelook_ecommerce.order_items` as items
ON users.id = items.user_id
WHERE
    status = 'Complete'
    AND EXTRACT(YEAR FROM items.created_at) = 2022
    AND EXTRACT(MONTH FROM items.created_at) BETWEEN 1 AND 9
    AND EXTRACT(QUARTER FROM items.created_at) < 4
GROUP BY Tahun, Bulan, Quarter
ORDER BY Bulan

''').to_dataframe()
user_transaksi

Tahun  Bulan  Quarter  Users
0   2022      1        1    649
1   2022      2        1    620
2   2022      3        1    767
3   2022      4        2    745
4   2022      5        2    825
5   2022      6        2    890
6   2022      7        3    920
7   2022      8        3    978
8   2022      9        3   1067

**INSIGHT**

Dalam menghitung user yang melakukan transaksi di tiap bulan, dataset yang digunakan adalah users dan order_items menggunakan fungsi JOIN, hanya saja pada aggregasi COUNT ini menggunakan DISTINCT agar user_id yang sama tidak dihitung secara berkala.

- User yang melakukan transaksi terendah tercatat pada bulan ke-2 sebesar 620 users, sementara users yang melakukan transaksi terbanyak tercatat pada bulan ke-9 sebanyak 1067.

- Dalam hal inipun memengaruhi total penjualan (USD) yang diperoleh, semakin banyak user yang bertransaksi maka semakin banyak pula pengahasilan total penjualan (USD). Dalam segi positif berarti promosi yang dilakukan oleh e-commerce ini dapat menarik perhatian users.

## **SOAL 4**

> Kategori produk apa saja yang paling banyak dibeli (status transaksi: Complete) di tiap bulannya selama Q1 sampai Q3 tahun 2022?

In [ ]:
# Melihat kategori produk yang paling banyak di beli
kategori_produk = client.query('''
SELECT Tahun, Bulan, Quarter, category AS Kategori, total AS Total_dibeli
FROM (
    SELECT
        Tahun, Bulan, Quarter, category, total,
        ROW_NUMBER() OVER (PARTITION BY Bulan ORDER BY total DESC) AS Partisi
    FROM (
        SELECT
            category,
            EXTRACT(Year FROM created_at) AS Tahun,
            EXTRACT(MONTH FROM created_at) as Bulan,
            EXTRACT(Quarter FROM created_at) AS Quarter,
            COUNT(products.id) as total
        FROM `bigquery-public-data.thelook_ecommerce.products` as products
        JOIN `bigquery-public-data.thelook_ecommerce.order_items` as items
        ON products.id = items.product_id
        WHERE
            status = 'Complete' AND
            created_at BETWEEN '2022-01-01' AND '2022-09-30'
        GROUP BY
            Tahun, Bulan, Quarter, category
        ) AS subs
) AS subs2
WHERE Partisi = 1
ORDER BY Bulan

''').to_dataframe()
kategori_produk

Tahun  Bulan  Quarter                       Kategori  Total_dibeli
0   2022      1        1                    Tops & Tees            67
1   2022      2        1                    Tops & Tees            72
2   2022      3        1                      Intimates            80
3   2022      4        2                      Intimates            87
4   2022      5        2                      Intimates            82
5   2022      6        2                      Intimates            97
6   2022      7        3  Fashion Hoodies & Sweatshirts           106
7   2022      8        3                      Intimates           121
8   2022      9        3                 Sleep & Lounge           108

**INSIGHT**

Fungsi yang digunakan untuk melihat kategori produk yang paling banyak di beli adalah window function dengan OVER partition by. Partisi ini di convert berdasarkan bulan dan kategori produk.

- Dalam kategori produk memiliki macam-macam ketersediaan, diantaranya Tops & Tees, Maternity, Sweater, Jeans, Outerwear & Coats, dan lain sebagainya. Dalam 3 quartil tersebut, produk yang paling sering menjadi tren atau paling sering di beli adalah Intimates selama 5 bulan.

- Agar kategori produk lain mengalami kenaikan tren, maka perlu melakukan promosi atau pemasaran lebih, melakukan inovasi terhadap kategori yang kurang tren agar menarik perhatian users.

## **SOAL 5**

> Kategori produk apa saja yang paling banyak mendapatkan pendapatan (status transaksi: Complete) di tiap bulannya selama Q1 sampai Q3 tahun 2022?

In [ ]:
# Melihat dan menghitung pendapatan paling banyak dari kategori produk
pendapatan_produk = client.query('''
SELECT
    Year,
    Month,
    Quarter,
    category AS Kategori,
    total_sales AS Total_Pendapatan
FROM (SELECT *,
            DENSE_RANK() OVER (PARTITION BY Month ORDER BY total_sales DESC) as rank
      FROM (SELECT
              EXTRACT(Year FROM created_at) AS Year,
              EXTRACT(MONTH FROM created_at) AS Month,
              EXTRACT(Quarter FROM created_at) AS Quarter,
              products.category,
              SUM(sale_price) AS total_sales,
            FROM `bigquery-public-data.thelook_ecommerce.order_items` AS items
            JOIN `bigquery-public-data.thelook_ecommerce.products` AS products
            ON items.product_id = products.id
            WHERE status='Complete'
                  AND EXTRACT(YEAR FROM created_at) = 2022
                  AND EXTRACT(MONTH FROM created_at) BETWEEN 1 AND 9
                  AND EXTRACT(QUARTER FROM created_at) < 4
            GROUP BY Year, Month, Quarter, products.category)
                     )
WHERE rank=1
ORDER BY Month''') \
.to_dataframe()
pendapatan_produk

Year  Month  Quarter           Kategori  Total_Pendapatan
0  2022      1        1  Outerwear & Coats       8672.089972
1  2022      2        1  Outerwear & Coats       8548.470037
2  2022      3        1  Outerwear & Coats       7597.950020
3  2022      4        2  Outerwear & Coats       9975.090000
4  2022      5        2              Jeans       8537.620043
5  2022      6        2  Outerwear & Coats      10529.359953
6  2022      7        3              Jeans       9295.780005
7  2022      8        3  Outerwear & Coats       8164.250017
8  2022      9        3  Outerwear & Coats       9633.389957

**INSIGHT**

Dalam kategori produk yang paling banyak dibeli rata-rata tiap bulan memang Intimates, namun untuk produk yang paling banyak menghasilkan pendapatan di tiap bulannya bukan Intimates, melainkan Outerwear & Coats pada bulan ke- 1, 2, 3, 4, 6, 8, 9, dan Jeans pada bulan ke 5 dan 7. Hal ini bisa saja terjadi karena harga masing-masing kategori berbeda.Dalam kurun waktu 9 bulan tersebut, total pendapatan tertinggi adalah bulan 6 dengan kategori Outerwear & Coats	yang menghasilkan total pendapatan sebesar 10529.359953.

## **SOAL 6**

> Dibandingkan dengan jumlah transaksi dan total penjualan, mana yang paling berkaitan dengan jumlah user yang melakukan transaksi? Apa analisis yang dapat kamu jelaskan?

In [ ]:
import pandas as pd
result = pd.concat([user_transaksi, Jumlah_transaksi, total_penjualan])
result

Tahun  Bulan  Quarter  Users  Jumlah_Transaksi  Total_penjualan
0   2022      1        1    649              <NA>              NaN
1   2022      2        1    620              <NA>              NaN
2   2022      3        1    767              <NA>              NaN
3   2022      4        2    745              <NA>              NaN
4   2022      5        2    825              <NA>              NaN
5   2022      6        2    890              <NA>              NaN
6   2022      7        3    920              <NA>              NaN
7   2022      8        3    978              <NA>              NaN
8   2022      9        3   1067              <NA>              NaN
0   2022      1        1   <NA>               917              NaN
1   2022      2        1   <NA>               902              NaN
2   2022      3        1   <NA>              1105              NaN
3   2022      4        2   <NA>              1053              NaN
4   2022      5        2   <NA>              1132              NaN
5   2022      6        2   <NA>              1334              NaN
6   2022      7        3   <NA>              1384              NaN
7   2022      8        3   <NA>              1386              NaN
8   2022      9        3   <NA>              1483              NaN
0   2022      1        1   <NA>              <NA>     1.179685e+06
1   2022      2        1   <NA>              <NA>     1.117165e+06
2   2022      3        1   <NA>              <NA>     1.301672e+06
3   2022      4        2   <NA>              <NA>     1.271810e+06
4   2022      5        2   <NA>              <NA>     1.434485e+06
5   2022      6        2   <NA>              <NA>     1.555542e+06
6   2022      7        3   <NA>              <NA>     1.558493e+06
7   2022      8        3   <NA>              <NA>     1.655541e+06
8   2022      9        3   <NA>              <NA>     1.775233e+06

**INSIGHT**

Jumlah user berdekatan dengan jumlah transaksi karena bisa melihat berapa total user yang melakukan transaksi berhasil atau gagal, apabisa berhasil masuk ke kategori jumlah user sehingga akan bertambah, namun jika gagal atau return otomatis jumlah users akan berkurang.

# **Poin Analisis Atau Kesimpulan**

Jumlah users yang berstatus complete dari quartil 1 sampai 3 mengalami kenaikan. Hal ini dapat mempengaruhi total penghasilan yang didapatkan dari masing-masing quartil.

Mengenai tren produk, kategori Fashion Hoodies & Sweatshirts serta Intimates memiliki jumlah terbanyak dari Q1 sampai Q3 di tahun 2022 yaitu sebanyak 3 bulan. Akan tetapi kategori Outerwear & Coats memiliki rata-rata pendapatan terbanyak dari Q1 sampai Q3 di tahun 2022.

Untuk mencapai peningkatan tren dalam kategori lain, diperlukan analisis pasar dan peminatan, peluang pasar, inovasi, dan strategi pemasaran.

Sehingga berdasarkan analisis tersebut, perusahaan e-commerce menunjukkan pertumbuhan yang positif dala hal transaksi, jumlah user yang bertransaksi, dan GMV (Gross Merchandise Value) selama 3 quartil terakhir. Namun, keputusan investor untuk menyuntikkan dana baru ke perusahaan tidak hanya didasarkan pada data historis, tetapi juga melihat prospek kedepan.

# **Poin Pertanyaan**

1. Apakah problem statement yang kamu definisikan di awal dapat terukur ketercapaiannya?

> Jawab : Menurut data yang telah dihitung, The Look ecommerce dapat mencapai tujuan yang diinginkan karena Q1 sampai Q3 pada tahun 2022 mengalami kenaikan. Selain itu, pendukung hal lainnya juga diperlukan agar kinerja penjualan di platform e-commerce mengalami peningkatan users yang berstatus Complete.

2. Berdasarkan hasil analisis yang sudah dilakukan dari 6 penjabaran di atas, jika CEO perusahaanmu ingin menargetkan pendapatan di awal kuartal 4 harus mencapai $250000, apakah masuk akal?

> Jawab : Menurut data pendapatan perbulan pada soal nomor 2, masing-masing quartil memang mengalami peningkatan namun peningkatan tersebut tidak meningkat secara drastis. Sehingga apabila CEO ingin di awal kuartal 4 mencapai 250000 USD sepertinya kurang masuk akal. Karena rentang hasil pendapatan yang di peroleh pada akhir kuartal 3 hanya mencapai 185358 USD.

3. CEO kamu menargetkan di kuartal 4 ada investor yang dapat menyuntikan dana ke perusahaanmu, dimana investor akan melihat GMV selama 3 kuartal terakhir serta prospek kedepan minimal di kuartal 4 akan seperti apa. Berikan informasi kepada CEO mu berdasarkan trend transaksi, jumlah user yang bertransaksi, dan GMV 3 kuartal terakhir, apakah perusahaanmu layak atau tidak mendapatkan investor baru!

 - Trend transaksi : dalam q1 hingga q3 pada tahun 2022 mengalami peningkatan dalam jumlah transaksi. Dikarenakan terjadi peningkatan yang signifikan, ini dapat menjadi indikator pertumbuhan yang positif dan menarik bagi investor.

 - Jumlah user yang bertransaksi : jumlah pengguna yang berpartisipasi dalam transaksi selama 3 quartil jumlahnya meningkat, hal ini menunjukkan adanya minat dan keterlibatan yang tinggi dari users serta dapat menjadi faktor menarik bagi investor.

 - Gross Merchandise Value (GMV) : GMV meningkat secara konsisten, hal ini akan memberikan keyakinan kepada investor bahwa perusahaan memiliki potensi pertumbuhan yang baik.

 - Dengan menganalisis data trend transaksi, jumlah user yang bertransaksi, dan GMV selama 3 quartil, maka perusahaan layak mendapat investor baru.